packages/runtime/datastore-definitions/src/jsonable.ts

```diff
@@ -3,19 +3,54 @@
  * Licensed under the MIT License.
  */
 
+/**
+ * Type constraint for types that are likely serializable as JSON or have a custom
+ * alternate type.
+ *
+ * @remarks
+ * Use `JsonableTypeWith<never>` for just JSON serializable types.
+ * See {@link Jsonable} for serialization pitfalls.
+ */
+export type JsonableTypeWith<T> =
+	| undefined
+	| null
+	| boolean
+	| number
+	| string
+	| T
+	| Internal_InterfaceOfJsonableTypesWith<T>
+	| ArrayLike<JsonableTypeWith<T>>;
+
+/**
+ * @remarks
+ * This type is a kludge and not intended for general use.
+ *
+ * @privateRemarks
+ * Internal type testing for compatibility uses TypeOnly filter which cannot handle recursive "pure" types.
+ * This interface along with ArrayLike above avoids pure type recursion issues.
+ * The TypeOnly filter is not useful for {@link JsonableTypeWith}; so, if type testing improves, this can be removed.
+ */
+export interface Internal_InterfaceOfJsonableTypesWith<T> {
+	[index: string | number]: JsonableTypeWith<T>;
```

Comment:
@CraigMacomber - I assume this is why the new 'Jsonable<T>' has trouble with interfaces as opposed to types.

An interface won't pattern match '{ [index: string]: ... }' because interfaces do not include '[index: string]: ...' (per https://github.com/microsoft/TypeScript/issues/15300#issuecomment-657218214, as Jason notes above.)

This makes me wonder if we want the old Jsonable<T> (but disallow T=any), plus the new JsonableType.

But it would be better if we could find a fix for interfaces...  Any thoughts?

Reply:
This is not why `Jsonable` has trouble with interfaces. If this was type alias, the same issue would apply.
This workaround is for the type test system that wraps types with `TypeOnly` even when that isn't helpful... and in the case of recursive types is problematic.